### ISSUE TITLE
rustc shell completions broken with recent nightly Rust

### ISSUE THREAD

**Issue by Nilstrieb** [fish-shell/fish-shell#10379]
I tried the following versions:
- fish 3.7.0 from nixpkgs on NixOS
- fish 3.7.0 from Ubuntu launchpad in a fresh `docker.io/ubuntu` container
- HEAD from source on Ubuntu decf99f71b2fb781013750898a33cd43847317f2 in a fresh `docker.io/rust`

Fish implements rustc completions by asking rustc to print its options. When `nightly-2024-03-16` Rust is installed (`rustup toolchain install nightly`), completing `rustc` (just `$ rustc <TAB>` is enough) crashes when trying to what looks like parsing rustc options. 

```
string split: - hashes of green query instances: unknown option

/usr/local/share/fish/completions/rustc.fish (line 1): 
string split -m 1 ' ' $line
^
in command substitution
	called on line 43 of file /usr/local/share/fish/completions/rustc.fish
from sourcing file /usr/local/share/fish/completions/rustc.fish

(Type 'help string' for related documentation)
string split: - hash collisions of query keys: unknown option

/usr/local/share/fish/completions/rustc.fish (line 1): 
string split -m 1 ' ' $line
^
in command substitution
	called on line 43 of file /usr/local/share/fish/completions/rustc.fish
from sourcing file /usr/local/share/fish/completions/rustc.fish

(Type 'help string' for related documentation)
```

I think the option that's causing this is `-Zincremental-verify-ich`, which prints this in `-Zhelp`:
```
    -Z                        incremental-verify-ich=val -- verify extended properties for incr. comp. (default: no):
        - hashes of green query instances
        - hash collisions of query keys
```

Those dashes are probably confusing it :D

**Comment by ian-h-chamberlain** [comment]
I tried out the fix implemented in 0ff1e50a33cc494ee2cd5f1490008e3e3e94e5df and while it prevents the error message in this issue, I think the completions are still partially incorrect:

```console
$ rustc -Z 
-Z(default:                                                                (`split`))
-Zall  (options: none, all, default, generic-activity, query-provider, query-cache-…)
-Zallow-features=  (only allow the listed language features to be enabled in code (…)
-Zalways-encode-mir=  (encode MIR of all functions into the crate metadata (default…)
-Zasm-comments=  (generate comments into the assembly (may change behavior) (defaul…)
-Zassert-incr-state=  (assert that the incremental cache is in given state: either …)
-Zassume-incomplete-release=  (make cfg(version) treat the current version as incom…)
-Zbinary-dep-depinfo=  (include artifacts (sysroot, crate dependencies) used during…)
-Zbox-noalias=                         (emit noalias metadata for box (default: yes))
-Zbranch-protection=  (set options for branch target identification and pointer aut…)
-Zby                                                                     (the linker)
-Zcf-protection=                    (instrument control-flow architecture protection)
…and 216 more rows
```

I think this is due to the help having multiple lines in certain cases, mainly if there are multiple choices for the right-hand-side of an option, e.g.
```
    -Z                              split-dwarf-kind=val -- split dwarf variant (only if -Csplit-debuginfo is enabled and on relevant platform)
        (default: `split`)

        `split`: sections which do not require relocation are written into a DWARF object (`.dwo`)
                 file which is ignored by the linker
        `single`: sections which do not require relocation are written into object file but ignored
                  by the linker
```

I'm not sure if it makes sense to include the extra lines in the description, since it would probably get cut off anyway, so maybe just filtering out any lines not containing `-Z` would be sufficient.

Not sure if this is worth a new issue or this should be reopened?

**Comment by zanchey** [comment]
I'll create a new issue.
